spiner/transformations.hpp

```diff
@@ -0,0 +1,60 @@
+#ifndef _SPINER_TRANSFORM_HPP_
+#define _SPINER_TRANSFORM_HPP_
+//======================================================================
+// © (or copyright) 2019-2021. Triad National Security, LLC. All rights
+// reserved.  This program was produced under U.S. Government contract
+// 89233218CNA000001 for Los Alamos National Laboratory (LANL), which is
+// operated by Triad National Security, LLC for the U.S.  Department of
+// Energy/National Nuclear Security Administration. All rights in the
+// program are reserved by Triad National Security, LLC, and the
+// U.S. Department of Energy/National Nuclear Security
+// Administration. The Government is granted for itself and others acting
+// on its behalf a nonexclusive, paid-up, irrevocable worldwide license
+// in this material to reproduce, prepare derivative works, distribute
+// copies to the public, perform publicly and display publicly, and to
+// permit others to do so.
+//======================================================================
+
+#include "ports-of-call/portability.hpp"
+
+#include <cmath>
+#include <limits>
+
+namespace Spiner {
+
+// Note on notation:
+// -- "real" space is called x
+// -- "transformed" space is called u
+// -- u = forward(x)
+// -- x = reverse(u)
+
+// linear transformation (aka no-op): y = x
+struct TransformLinear {
+  template <typename T>
+  PORTABLE_INLINE_FUNCTION static T forward(const T x) {
+    return x;
+  }
+  template <typename T>
+  PORTABLE_INLINE_FUNCTION static T reverse(const T u) {
+    return u;
+  }
+};
+
+// logarithmic transformation: y = log(x + small)
+struct TransformLogarithmic {
+  template <typename T>
+  PORTABLE_INLINE_FUNCTION static T forward(const T x) {
+    return std::log(x + std::numeric_limits<T>::denorm_min());
```

Comment:
```suggestion
   return std::log(std::abs(x) + std::numeric_limits<T>::denorm_min());
```
I'm also not sure `denorm_min` is appropriate here. Not obviously inappropriate either though.

Reply:
I figured that `denorm_min` is the smallest representable value, and therefore the least likely to skew the results for any _actual_ data value.  But if you think something else is more appropriate, that's fine.

Reply:
I think in this case the value proposition of denormal_min is not great enough to justify it's use over `min` https://stackoverflow.com/questions/54937154/why-are-denormal-floating-point-values-slower-to-handle

Reply:
I didn't know that about denormalized values, thanks @dholladay00!  I think that's an excellent argument for `min` instead of `denorm_min`, so I'll make the change.

Reply:
I changed from `denorm_min` to `min`, but that exposed a problem where the round-trip is wrong for the critical value of zero (that is, `reverse(forward(0))` is not equal to 0).  Because of that I (a) updated the tests a little, and (b) modified the transformation so that the epsilon values are asymmetric in a way that gets `reverse(forward(0))` _exactly_ correct.  Unfortunately, `std::exp` and `std::log` are not `constexpr` until C++26, so the best I could do is pre-compute and hard-code the values for `float` and `double`, but fall back to a symmetric epsilon for all other types.

This is all a little unexpected, so please take a look and let me know what you think, @dholladay00 and @Yurlungur .

Reply:
This is totally reasonable and I play a similar game elsewhere.